drivers/dma/dma_iproc_pax_v2.c

```diff
@@ -0,0 +1,940 @@
+	struct device *pcidev;
+	struct dma_iproc_pax_write_sync_data sync_rd, *recv, *sent;
+	uint64_t pci_addr;
+	uint32_t *pci32, *axi32;
+	int loop, ret;
+	uint32_t zero_init = 0;
+
+	pcidev = device_get_binding(cfg->pcie_dev_name);
+	if (!pcidev) {
+		LOG_ERR("Cannot get pcie device\n");
+		return -EINVAL;
+	}
+
+	recv = &sync_rd;
+	sent = &(ring->curr.sync_data);
+	/* form host pci sync address */
+	pci32 = (uint32_t *)&pci_addr;
+	pci32[0] = ring->sync_pci.addr_lo;
+	pci32[1] = ring->sync_pci.addr_hi;
+	axi32 = (uint32_t *)&sync_rd;
+
+	for (loop = 0; loop < MAX_SYNC_WAIT; loop++) {
+
+		ret = pcie_ep_xfer_data_memcpy(pcidev, pci_addr,
+					       (uintptr_t *)axi32, 4,
+					       PCIE_OB_LOWMEM, HOST_TO_DEVICE);
+
+		if (memcmp((void *)recv, (void *)sent, 4) == 0) {
+			/* clear the sync word */
+			ret = pcie_ep_xfer_data_memcpy(pcidev, pci_addr,
+						       (uintptr_t *)&zero_init,
+						       4, PCIE_OB_LOWMEM,
+						       DEVICE_TO_HOST);
+			dma_mb();
+			ret = 0;
+			break;
+		}
+	}
+	if (loop == MAX_SYNC_WAIT) {
+		LOG_ERR("[ring %d]: not recvd write sync!\n", ring->idx);
+		ret = -ETIMEDOUT;
+	}
+
+	return ret;
+}
+
+static int process_cmpl_event(struct dma_iproc_pax_data *pd,
+			      enum ring_idx idx, uint32_t pl_len)
+{
+	uint32_t wr_offs, rd_offs, ret = 0;
+	struct dma_iproc_pax_rm_ring_data *ring = &(pd->ring[idx]);
+	struct cmpl_pkt *c;
+	uint32_t is_outstanding;
+
+	/* cmpl read offset, unprocessed cmpl location */
+	rd_offs = ring->curr.cmpl_rd_offs;
+
+	wr_offs = sys_read32(RM_RING_REG(pd, idx,
+					 RING_CMPL_WRITE_PTR));
+
+	/* Update read ptr to "processed" */
+	ring->curr.cmpl_rd_offs = wr_offs;
+
+	/*
+	 * Ensure consistency of completion descriptor
+	 * The completion desc is updated by RM via AXI stream
+	 * CPU need to ensure the memory operations are completed
+	 * before reading cmpl area, by a "dsb"
+	 * If Dcache enabled, need to invalidate the cachelines to
+	 * read updated cmpl desc. The cache API also issues dsb.
+	 */
+	dma_mb();
+
+	/* Decode cmpl pkt id to verify */
+	c = (struct cmpl_pkt *)((uint32_t)ring->cmpl +
+	    CMPL_DESC_SIZE * CURR_CMPL_IDX(wr_offs));
+
+	LOG_DBG("RING%d WR_PTR:%d opq:%d, rm_status:%x dma_status:%x\n",
+		idx, wr_offs, c->opq, c->rm_status, c->dma_status);
+
+	is_outstanding = sys_read32(RM_RING_REG(pd, idx,
+						RING_NUM_REQ_OUTSTAND));
+	if ((ring->curr.opq != c->opq) && (is_outstanding != 0)) {
+		LOG_ERR("RING%d: pkt id should be %d, rcvd %d outst=%d\n",
+			idx, ring->curr.opq, c->opq, is_outstanding);
+		ret = -EIO;
+	}
+	/* check for completion AE timeout */
+	if (c->rm_status == RM_COMPLETION_AE_TIMEOUT) {
+		LOG_ERR("RING%d WR_PTR:%d rm_status:%x AE Timeout!\n",
+			idx, wr_offs, c->rm_status);
+		/* TBD: Issue full card reset to restore operations */
+		LOG_ERR("Needs Card Reset to recover!\n");
+		ret = -ETIMEDOUT;
+	}
+
+	ring->dma_callback(ring->callback_arg, idx, ret);
+
+	return ret;
+}
+
+#ifdef CONFIG_DMA_IPROC_PAX_POLL_MODE
```

Comment:
Doing this as an if/else, but doing completely different things in the two blocks, can be confusing. It's easier to follow the code if you do two independent ifdefs.

Reply:
ack, will fix it.